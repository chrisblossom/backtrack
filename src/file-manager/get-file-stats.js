/* @flow */

import type { ParsedFiles, FileManagerStats } from '../types.js';

function getFileStats(parsedFiles: ParsedFiles): FileManagerStats {
    const fileHashes = parsedFiles.dest.files.reduce((acc, destFile) => {
        const destFileIndex = parsedFiles.dest.files.indexOf(destFile);

        /**
         * Use source file's hash to allow for 'allowChanges' option and guard against file system errors
         *
         * destHash will equal srcHash because the copy_files task is done prior to stats
         */
        const srcFile = parsedFiles.src.files[destFileIndex];
        const srcHash = parsedFiles.src.hash[srcFile];

        return {
            ...acc,
            [destFile]: srcHash,
        };
    }, {});

    const stats = {};
    if (Object.keys(fileHashes).length !== 0) {
        stats.files = fileHashes;
    }

    if (parsedFiles.makeDirs.length !== 0) {
        stats.directories = parsedFiles.makeDirs;
    }

    return stats;
}

export { getFileStats };
