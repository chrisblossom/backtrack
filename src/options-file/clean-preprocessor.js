/* @flow */

import { toArray } from '../utils/object-utils';
import type { Clean, NormalizedClean } from '../types.js';

// eslint-disable-next-line flowtype/require-exact-type
type Args = {
    value?: Clean | $ReadOnlyArray<Clean>,
};

function cleanPreprocessor({ value }: Args = {}): $ReadOnlyArray<
    NormalizedClean,
> {
    const allToArray = toArray(value);

    return allToArray.map((arg: Clean) => {
        return {
            del: toArray(arg.del),
            makeDirs: toArray(arg.makeDirs),
        };
    });
}

export { cleanPreprocessor };
