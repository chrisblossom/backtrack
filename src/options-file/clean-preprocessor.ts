import path from 'path';
import { toArray } from '../utils/object-utils';
import { buildPath, rootPath } from '../config/paths';
import { Clean, NormalizedClean } from '../types';

interface Args {
	value?: Clean | Clean[];
}

function cleanPreprocessor({ value }: Args = {}): NormalizedClean[] {
	const allToArray = toArray(value);

	return allToArray.map((arg: Clean) => {
		const makeDirs = toArray(arg.makeDirs).map((makeDir) => {
			const resolvedMakeDir = path.resolve(buildPath, makeDir);
			return resolvedMakeDir;
		});

		const copy = toArray(arg.copy).map((copyArgs) => {
			const resolvedSrc = path.resolve(rootPath, copyArgs.src);
			const resolvedDest = path.resolve(buildPath, copyArgs.dest);

			const result = {
				src: resolvedSrc,
				dest: resolvedDest,
				hash: copyArgs.hash,
			};

			if (result.hash === undefined) {
				delete result.hash;
			}

			return result;
		});

		return {
			del: toArray(arg.del),
			makeDirs,
			copy,
		};
	});
}

export { cleanPreprocessor };
