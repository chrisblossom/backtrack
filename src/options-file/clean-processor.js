/* @flow */

import path from 'path';
import { buildPath } from '../config/paths';

import type { NormalizedClean } from '../types.js';

/**
 * Patterns are only allowed inside the buildPath.
 */
const base = {
    del: [],
    makeDirs: [],
};

// eslint-disable-next-line flowtype/require-exact-type
type Args = {
    value: $ReadOnlyArray<NormalizedClean>,
    current?: NormalizedClean,
};

function cleanProcessor({ value, current = base }: Args) {
    return value.reduce(
        (acc, arg) => {
            arg.del.forEach((pattern) => {
                /**
                 * remove duplicates
                 */
                if (acc.del.includes(pattern) === false) {
                    acc.del.push(pattern);
                }
            });

            arg.makeDirs.forEach((dir) => {
                const resolvedDir = path.resolve(buildPath, dir);

                /**
                 * remove duplicates
                 */
                if (acc.makeDirs.includes(resolvedDir) === false) {
                    acc.makeDirs.push(resolvedDir);
                }
            });

            return acc;
        },
        {
            del: [...current.del],
            makeDirs: [...current.makeDirs],
        },
    );
}

export { cleanProcessor };
