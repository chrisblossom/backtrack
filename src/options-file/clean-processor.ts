import { NormalizedClean } from '../types';

/**
 * Patterns are only allowed inside the buildPath.
 */
const base = {
	del: [],
	makeDirs: [],
	copy: [],
};

interface Args {
	value: NormalizedClean[];
	current?: NormalizedClean;
}

function cleanProcessor({ value, current = base }: Args) {
	return value.reduce(
		(acc, arg) => {
			arg.del.forEach((pattern) => {
				/**
				 * remove duplicates
				 */
				if (acc.del.includes(pattern) === false) {
					acc.del.push(pattern);
				}
			});

			arg.makeDirs.forEach((dir) => {
				/**
				 * remove duplicates
				 */
				if (acc.makeDirs.includes(dir) === false) {
					acc.makeDirs.push(dir);
				}
			});

			arg.copy.forEach((copy) => {
				/**
				 * remove duplicates
				 */
				const isDuplicate = !!acc.copy.find((copyArg) => {
					return (
						copyArg.src === copy.src && copyArg.dest === copy.dest
					);
				});

				if (isDuplicate === false) {
					acc.copy.push(copy);
				}
			});

			return acc;
		},
		{
			del: [...current.del],
			makeDirs: [...current.makeDirs],
			copy: [...current.copy],
		},
	);
}

export { cleanProcessor };
