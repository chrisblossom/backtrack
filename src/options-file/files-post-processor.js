/* @flow */

import path from 'path';
import { parseFilePath } from '../utils/parse-file-path';
import { normalizePathname } from '../utils/normalize-pathname';
import { toArray } from '../utils/object-utils';

import type { FileManager, ParsedFiles } from '../types.js';

function shouldSkip(pathname, skipFiles) {
    const split = pathname.split(path.sep);

    let currentPath;
    for (const sub of split) {
        currentPath = currentPath ? currentPath + path.sep + sub : sub;

        if (skipFiles.includes(currentPath)) {
            return true;
        }
    }

    return false;
}

function splitSections(copyFiles) {
    return copyFiles.reduce(
        (acc, file) => {
            if (file.src && file.dest) {
                return {
                    ...acc,
                    copyFiles: [...acc.copyFiles, file],
                };
            }

            // $FlowIssue
            const { makeDirs = [], ...options } = file;

            return {
                ...acc,
                options: [...acc.options, options],
                makeDirs: [...acc.makeDirs, ...makeDirs],
            };
        },
        { copyFiles: [], options: [], makeDirs: [] },
    );
}

function parseOptions(options) {
    return options.reduce(
        (acc, fileOptions) => {
            if (fileOptions.skip) {
                toArray(fileOptions.skip).forEach((relativePath) => {
                    const normalizedFile = normalizePathname(relativePath);

                    acc.skip.push(normalizedFile);
                });
            }

            if (typeof fileOptions.allowChanges === 'boolean') {
                acc.allowChangesAll = fileOptions.allowChanges;

                return acc;
            }

            toArray(fileOptions.allowChanges).forEach((file) => {
                const normalizedFile = normalizePathname(file);

                acc.allowChanges.push(normalizedFile);
            });

            return acc;
        },
        {
            allowChangesAll: false,
            allowChanges: [],
            skip: [],
        },
    );
}

function getMakeDirs(makeDirs, files, skipFiles) {
    const filePaths = files
        .map((filePath) => {
            const { dir } = path.parse(filePath);

            return dir;
        })
        .filter(Boolean);

    const result = [...filePaths, ...makeDirs]
        // eslint-disable-next-line arrow-body-style
        .map((dir) => normalizePathname(dir))
        .sort()
        .reduceRight((acc, dir) => {
            /**
             * remove skipped
             */
            const skip = shouldSkip(dir, skipFiles);

            if (skip === true) {
                return acc;
            }

            /**
             * removes redundant directories
             * ['src', 'dist', 'dist/static'] -> ['src', 'dist/static']
             */
            const previous = acc[0];
            if (previous) {
                const trimmedPrevious = previous.substring(0, dir.length);

                if (trimmedPrevious === dir) {
                    return acc;
                }
            }

            return [dir, ...acc];
        }, []);

    return result;
}

// eslint-disable-next-line flowtype/require-exact-type
type Args = {
    value: FileManager,
};

/**
 * Normally this would be a normal processor, but we need to have all results up front
 * to determine what to exclude etc. I'm sure the function could be rewritten but this works as is.
 */
function filesPostProcessor({ value = [] }: Args = {}): ParsedFiles {
    const result = {
        src: {
            files: [],
            absolute: {},
            hash: {},
        },
        dest: {
            files: [],
            absolute: {},
            hash: {},
            allowChanges: {},
        },
        makeDirs: [],
    };

    const sections = splitSections(value);
    const options = parseOptions(sections.options);

    for (const file of sections.copyFiles) {
        const src = parseFilePath(file.src);
        const dest = parseFilePath(file.dest);

        const skip = shouldSkip(dest.relative, options.skip);
        if (skip === false) {
            const srcId = src.relative;
            result.src.files.push(srcId);
            result.src.absolute[srcId] = src.absolute;
            result.src.hash[srcId] = src.hash;

            const destId = dest.relative;
            result.dest.files.push(destId);
            result.dest.absolute[destId] = dest.absolute;
            result.dest.hash[destId] = dest.hash;

            const allowChanges =
                options.allowChanges.includes(destId) || !!file.allowChanges;

            result.dest.allowChanges[destId] =
                options.allowChangesAll || allowChanges;
        }
    }

    result.makeDirs = getMakeDirs(
        sections.makeDirs,
        result.dest.files,
        options.skip,
    );

    return result;
}

export { filesPostProcessor };
