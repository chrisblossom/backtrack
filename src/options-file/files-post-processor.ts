import path from 'path';
import { parseFilePath } from '../utils/parse-file-path';
import { normalizePathname } from '../utils/normalize-pathname';
import { toArray } from '../utils/object-utils';

import {
    FileManager,
    CopyFile,
    isCopyFile,
    CopyFileOptions,
    ParsedFiles,
} from '../types';

function shouldSkip(pathname: string, skipFiles: Array<string>) {
    const split = pathname.split(path.sep);

    let currentPath;
    for (const sub of split) {
        currentPath = currentPath ? currentPath + path.sep + sub : sub;

        if (skipFiles.includes(currentPath)) {
            return true;
        }
    }

    return false;
}

type SplitSections = {
    copyFiles: CopyFile[];
    options: CopyFileOptions[];
    makeDirs: string[];
};

function splitSections(copyFiles: FileManager): SplitSections {
    type SectionsAcc = {
        copyFiles: CopyFile[];
        options: CopyFileOptions[];
        makeDirs: string[];
    };

    const sections = copyFiles.reduceRight(
        (acc: SectionsAcc, file) => {
            if (isCopyFile(file)) {
                const isDuplicate = acc.copyFiles.some((accFile) => {
                    return file.dest === accFile.dest;
                });

                /**
                 * Remove duplicates
                 */
                if (isDuplicate) {
                    return acc;
                }

                return {
                    ...acc,
                    copyFiles: [file, ...acc.copyFiles],
                };
            }

            const { makeDirs = [], ...options } = file;

            return {
                ...acc,
                options: [options, ...acc.options],
                makeDirs: [...makeDirs, ...acc.makeDirs],
            };
        },
        { copyFiles: [], options: [], makeDirs: [] },
    );

    return sections;
}

function parseOptions(options: CopyFileOptions[]) {
    type ParseOptionsAcc = {
        allowChangesAll: boolean;
        ignoreUpdatesAll: boolean;
        allowChanges: string[];
        ignoreUpdates: string[];
        skip: string[];
    };

    const result = options.reduce(
        (acc: ParseOptionsAcc, fileOptions) => {
            if (fileOptions.skip) {
                toArray(fileOptions.skip).forEach((relativePath) => {
                    const normalizedFile = normalizePathname(relativePath);

                    acc.skip.push(normalizedFile);
                });
            }

            if (typeof fileOptions.allowChanges === 'boolean') {
                acc.allowChangesAll = fileOptions.allowChanges;

                return acc;
            }

            toArray(fileOptions.allowChanges).forEach((file) => {
                const normalizedFile = normalizePathname(file);

                acc.allowChanges.push(normalizedFile);
            });

            if (typeof fileOptions.ignoreUpdates === 'boolean') {
                acc.ignoreUpdatesAll = fileOptions.ignoreUpdates;

                return acc;
            }

            toArray(fileOptions.ignoreUpdates).forEach((file) => {
                const normalizedFile = normalizePathname(file);

                acc.ignoreUpdates.push(normalizedFile);
            });

            return acc;
        },
        {
            allowChangesAll: false,
            ignoreUpdatesAll: false,
            allowChanges: [],
            ignoreUpdates: [],
            skip: [],
        },
    );

    return result;
}

function getMakeDirs(makeDirs: string[], files: string[], skipFiles: string[]) {
    const filePaths = files
        .map((filePath) => {
            const { dir } = path.parse(filePath);

            return dir;
        })
        .filter(Boolean);

    const result = [...filePaths, ...makeDirs]
        // eslint-disable-next-line arrow-body-style
        .map((dir) => normalizePathname(dir))
        .sort()
        .reduceRight((acc: string[], dir) => {
            /**
             * remove skipped
             */
            const skip = shouldSkip(dir, skipFiles);

            if (skip === true) {
                return acc;
            }

            /**
             * removes redundant directories
             * ['src', 'dist', 'dist/static'] -> ['src', 'dist/static']
             */
            const previous = acc[0];
            if (previous) {
                const trimmedPrevious = previous.substring(0, dir.length);

                if (trimmedPrevious === dir) {
                    return acc;
                }
            }

            return [dir, ...acc];
        }, []);

    return result;
}

type Args = {
    value?: FileManager;
};

/**
 * Normally this would be a normal processor, but we need to have all results up front
 * to determine what to exclude etc. I'm sure the function could be rewritten but this works as is.
 */
function filesPostProcessor({ value = [] }: Args = {}): ParsedFiles {
    const result: ParsedFiles = {
        src: {
            files: [],
            absolute: {},
            hash: {},
        },
        dest: {
            files: [],
            absolute: {},
            hash: {},
            allowChanges: {},
            ignoreUpdates: {},
        },
        makeDirs: [],
    };

    const sections = splitSections(value);
    const options = parseOptions(sections.options);

    for (const file of sections.copyFiles) {
        const src = parseFilePath(file.src);
        const dest = parseFilePath(file.dest);

        const skip = shouldSkip(dest.relative, options.skip);
        if (skip === false) {
            const srcId = src.relative;
            const destId = dest.relative;

            const ignoreUpdates =
                options.ignoreUpdatesAll ||
                options.ignoreUpdates.includes(destId) ||
                !!file.ignoreUpdates;

            result.src.files.push(srcId);
            result.src.absolute[srcId] = src.absolute;
            result.src.hash[srcId] =
                ignoreUpdates === false ? src.hash : 'ignoreUpdates';

            result.dest.files.push(destId);
            result.dest.absolute[destId] = dest.absolute;
            result.dest.hash[destId] = dest.hash;

            const allowChanges =
                ignoreUpdates ||
                options.allowChanges.includes(destId) ||
                !!file.allowChanges;

            result.dest.ignoreUpdates[destId] = ignoreUpdates;

            result.dest.allowChanges[destId] =
                options.ignoreUpdatesAll ||
                options.allowChangesAll ||
                allowChanges;
        }
    }

    result.makeDirs = getMakeDirs(
        sections.makeDirs,
        result.dest.files,
        options.skip,
    );

    return result;
}

export { filesPostProcessor };
