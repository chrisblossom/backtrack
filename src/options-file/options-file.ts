import { loadOptionsFile } from './load-options-file';
import { transformConfig } from './transform-config';
import { rootPath } from '../config/paths';

import { Lifecycles } from '../types';

function optionsFile(dirname: string = rootPath): Lifecycles {
    const baseConfig = loadOptionsFile(dirname);

    const config = transformConfig(baseConfig, dirname);

    return config;
}

export { optionsFile };
