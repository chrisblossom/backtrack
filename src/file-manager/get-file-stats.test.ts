import path from 'path';

require('./get-file-stats');
require('../options-file/files-post-processor');

const getFileStats = (files: any) =>
    require('./get-file-stats').getFileStats(files);
const filesParser = (args: any) =>
    require('../options-file/files-post-processor').filesPostProcessor(args);

describe('getFileStats', () => {
    const cwd = process.cwd();

    afterEach(() => {
        process.chdir(cwd);
    });

    test('returns file stats', () => {
        const dir = path.resolve(__dirname, '__sandbox__/stats1/');
        process.chdir(dir);

        const value = [
            {
                src: 'nested/other.js',
                dest: 'nested/other.js',
            },
            {
                src: 'nested/inside.js',
                dest: 'nested/two/inside.js',
            },
            {
                src: 'z-file2.js',
                dest: 'z-file2.js',
            },
            {
                src: 'file1.js',
                dest: 'file1.js',
            },
            {
                makeDirs: ['dist', 'src'],
            },
        ];

        const parsedFiles = filesParser({ value });

        const result = getFileStats(parsedFiles);

        expect(result).toMatchSnapshot();
    });

    test('removes files when none exist', () => {
        const dir = path.resolve(__dirname, '__sandbox__/stats1/');
        process.chdir(dir);

        const value = [
            {
                makeDirs: ['dist', 'src'],
            },
        ];

        const parsedFiles = filesParser({ value });

        const result = getFileStats(parsedFiles);

        expect(result).toMatchSnapshot();
    });

    test('removes makeDirs when none exist', () => {
        const dir = path.resolve(__dirname, '__sandbox__/stats1/');
        process.chdir(dir);

        const value = [
            {
                src: 'file1.js',
                dest: 'file1.js',
            },
        ];

        const parsedFiles = filesParser({ value });

        const result = getFileStats(parsedFiles);

        expect(result).toMatchSnapshot();
    });

    test('returns empty object when no stats', () => {
        const dir = path.resolve(__dirname, '__sandbox__/stats1/');
        process.chdir(dir);

        const value: Array<void> = [];

        const parsedFiles = filesParser({ value });

        const result = getFileStats(parsedFiles);

        expect(result).toMatchSnapshot();
    });
});
