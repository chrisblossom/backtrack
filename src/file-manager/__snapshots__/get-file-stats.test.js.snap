// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getFileStats removes files when none exist 1`] = `
Object {
  "directories": Array [
    "dist",
    "src",
  ],
}
`;

exports[`getFileStats removes makeDirs when none exist 1`] = `
Object {
  "files": Object {
    "file1.js": "971eb09e78cb5dca6dd4fba0284b996b624df850e0e27d2933280271d3a7b856",
  },
}
`;

exports[`getFileStats returns empty object when no stats 1`] = `Object {}`;

exports[`getFileStats returns file stats 1`] = `
Object {
  "directories": Array [
    "dist",
    "nested/two",
    "src",
  ],
  "files": Object {
    "file1.js": "971eb09e78cb5dca6dd4fba0284b996b624df850e0e27d2933280271d3a7b856",
    "nested/other.js": "44c09b273fa7e0ffcb87b76cf882a5c7cdeba79e80b8bc814d3909a2a58ad995",
    "nested/two/inside.js": "46234eada9c51bae5d7a671fbb939e6efd1dda8b524f2b8b015aa4bce7746c29",
    "z-file2.js": "447ad688b0d3b7449000199a9e766ee03965da1ab15a961756a90dcff3a810e3",
  },
}
`;
